src/main/java/org/folio/rest/support/ShelvingOrderUpdate.java

```diff
@@ -0,0 +1,366 @@
+
+      if (ar.failed()) {
+        log.info("Connection acquiring failure : {}", ar.cause().getMessage());
+        fetchAndUpdatePromise.fail("Connection acquiring failure : " + ar.cause().getMessage());
+
+      } else {
+        PgConnection connection = ar.result();
+        // Starts a new transaction (acquiring and updating of items performs in single transaction)
+        connection.begin()
+            // New transaction started
+            .compose(tx -> connection
+
+            // Acquiring current bunch of items
+            .preparedQuery(SQL_SELECT_ITEMS).execute(Tuple.of(rowsBunchSize))
+
+            .compose(fetchedRows ->
+                // Read items and fill them into data structures for further processing
+                aggregateRow2List(fetchedRows)
+
+                  // Calculate items shelving order
+                  .compose(ShelvingOrderUpdate.this::calculateShelvingOrder)
+
+                  // Prepare items update parameters
+                  .compose(ShelvingOrderUpdate.this::prepareItemsUpdate)
+
+                  // Updates of items
+                  .compose(listOfTuple -> updateItems(listOfTuple, connection))
+                    .onFailure(h -> {
+                      log.error("Failed updateItems execution: {}, transaction will be rolled back/", h.getMessage());
+                      tx.rollback();
+                      })
+
+                  // Finish transaction
+                  .compose(updatedRowsCount -> completeTransaction(tx, updatedRowsCount, fetchAndUpdatePromise))
+
+                  // Close connection
+                  .eventually(v -> connection.close()))
+                .onFailure(h -> {
+                  tx.rollback();
+                  log.error("Error occurred in fetchAndUpdatesItems: {}, rolled back transaction...", h.getMessage());
+                  fetchAndUpdatePromise.fail(h.getCause());
+                })
+            );
+      }
+    });
+
+    return fetchAndUpdatePromise.future();
+  }
+
+  /**
+   * Completes transaction and set common update promise with affected rows count
+   *
+   * @param tx
+   * @param updatedRowsCount
+   * @param itemsUpdatePromise
+   * @return
+   */
+  private Future<Void> completeTransaction(Transaction tx, Integer updatedRowsCount,
+      Promise<Integer> itemsUpdatePromise) {
+    log.info("Invoking of completeTransaction, updatedRows size: {}", updatedRowsCount);
+
+    Future<Void> commitFuture = tx.commit()
+        .onSuccess(h -> itemsUpdatePromise.complete(updatedRowsCount))
+        .onFailure(h -> itemsUpdatePromise.fail(h.getCause().getMessage()));
+
+    log.info("Finishing of \"completeTransaction\", itemsUpdatePromise: {}, commitFuture: {}", itemsUpdatePromise, commitFuture);
+    return Future.succeededFuture();
+  }
+
+  /**
+   * Read items and fill them into data structures for further processing
+   *
+   * @param rowSet
+   * @return
+   */
+  public Future<List<Item>> aggregateRow2List(RowSet<Row> rowSet) {
```

Comment:
No async operation is involved, this method should return `List<Item>`, no `Future` wrapping is needed.

Reply:
Done, thank you.

Reply:
Done, thank you.